iceoryx_utils/test/moduletests/test_concurrent_smart_lock.cpp

```diff
@@ -0,0 +1,358 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "iceoryx_utils/cxx/optional.hpp"
+#include "iceoryx_utils/cxx/vector.hpp"
+#include "iceoryx_utils/internal/concurrent/smart_lock.hpp"
+#include "iceoryx_utils/posix_wrapper/semaphore.hpp"
+#include "test.hpp"
+
+#include <atomic>
+#include <thread>
+
+using namespace ::testing;
+using namespace iox::concurrent;
+using namespace iox::cxx;
+
+namespace
+{
+class SmartLockTester
+{
+  public:
+    // just for clarity, we could also write ++m_a but we want
+    // to highlight the possible race condition here when we call this from
+    // multiple threads, therefore m_a = m_a + 1
+    SmartLockTester()
+    {
+    }
+
+    SmartLockTester(const int32_t a)
+        : m_a(a)
+    {
+    }
+
+    SmartLockTester(const SmartLockTester& rhs)
+        : m_a(rhs.m_a)
+    {
+        rhs.m_b = rhs.m_b + 1;
+    }
+
+    SmartLockTester(SmartLockTester&& rhs)
+        : m_a(rhs.m_a)
+    {
+        rhs.m_a = 0;
+        rhs.m_b = rhs.m_b + 1;
+    }
+
+    SmartLockTester& operator=(const SmartLockTester& rhs)
+    {
+        if (this != &rhs)
+        {
+            rhs.m_b = rhs.m_b + 1;
+            m_a = rhs.m_a;
+        }
+        return *this;
+    }
+
+    SmartLockTester& operator=(SmartLockTester&& rhs)
+    {
+        if (this != &rhs)
+        {
+            rhs.m_b = rhs.m_b + 1;
+            m_a = rhs.m_a;
+            rhs.m_a = 0;
+        }
+        return *this;
+    }
+
+    ~SmartLockTester()
+    {
+    }
+
+    int32_t getA() const
+    {
+        return m_a;
+    }
+
+    int32_t getB() const
+    {
+        return m_b;
+    }
+
+    void setA(const int32_t a)
+    {
+        m_a = a;
+    }
+
+    void incrementA()
+    {
+        m_a = m_a + 1;
+    }
+
+    void constIncrementA() const
+    {
+        m_a = m_a + 1;
+    }
+
+  private:
+    mutable int32_t m_a = 0;
+    mutable int32_t m_b = 0;
+};
+
+class smart_lock_test : public Test
+{
+  public:
+    void SetUp() override
+    {
+    }
+    void TearDown() override
+    {
+    }
+
+    void waitUntilThreadsHaveStarted(const uint64_t numberOfThreads)
+    {
+        ++m_numberOfThreadWaiter;
+        while (m_numberOfThreadWaiter.load() != numberOfThreads)
+        {
+        }
+    }
+
+    using SutType_t = smart_lock<SmartLockTester>;
+    optional<SutType_t> m_sut;
+    std::atomic<uint64_t> m_numberOfThreadWaiter{0U};
+};
+constexpr uint64_t NUMBER_OF_RUNS_PER_THREAD = 100000U;
+constexpr uint64_t NUMBER_OF_THREADS = 4;
+
+} // namespace
+
+//////////////////////////////////////
+// BEGIN single threaded api test
+//////////////////////////////////////
+
+TEST_F(smart_lock_test, defaultConstructionOfUnderlyingObjectWorks)
```

Comment:
Most of the tests have an upper case test name. We should not increase the inconsistency by adding new tests with lower case test names